cmd/pullrequest-init/fake_github.go

```diff
@@ -0,0 +1,171 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/google/go-github/github"
+	"github.com/gorilla/mux"
+)
+
+// key defines keys for associating data to PRs/issues in the fake server.
+type key struct {
+	owner string
+	repo  string
+	id    int64
+}
+
+// FakeGitHub is a fake GitHub server for use in tests.
+type FakeGitHub struct {
+	*mux.Router
+
+	pr       map[key]*github.PullRequest
+	comments map[key][]*github.IssueComment
+}
+
+// NewFakeGitHub returns a new FakeGitHub.
+func NewFakeGitHub() *FakeGitHub {
+	s := &FakeGitHub{
+		Router:   mux.NewRouter(),
+		pr:       make(map[key]*github.PullRequest),
+		comments: make(map[key][]*github.IssueComment),
+	}
+	s.HandleFunc("/repos/{owner}/{repo}/pulls/{number}", s.getPullRequest).Methods(http.MethodGet)
+	s.HandleFunc("/repos/{owner}/{repo}/issues/{number}/comments", s.getComments).Methods(http.MethodGet)
+	s.HandleFunc("/repos/{owner}/{repo}/issues/{number}/comments", s.createComment).Methods(http.MethodPost)
+	s.HandleFunc("/repos/{owner}/{repo}/issues/{number}/labels", s.updateLabels).Methods(http.MethodPut)
+
+	return s
+}
+
+// AddPullRequest adds the given pull request to the fake GitHub server.
+func (g *FakeGitHub) AddPullRequest(pr *github.PullRequest) {
+	key := key{
+		owner: pr.GetBase().GetUser().GetLogin(),
+		repo:  pr.GetBase().GetRepo().GetName(),
+		id:    pr.GetID(),
+	}
+	g.pr[key] = pr
+}
+
+// AddComment adds a comment to the fake GitHub server.
+func (g *FakeGitHub) AddComment(owner string, repo string, pr int64, comment *github.IssueComment) {
+	key := key{
+		owner: owner,
+		repo:  repo,
+		id:    pr,
+	}
+	c := g.comments[key]
+	if c == nil {
+		c = []*github.IssueComment{comment}
+	} else {
+		c = append(c, comment)
+	}
+	g.comments[key] = c
+}
+
+func (g *FakeGitHub) getPullRequest(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.ParseInt(mux.Vars(r)["number"], 10, 64)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	key := key{
+		owner: mux.Vars(r)["owner"],
+		repo:  mux.Vars(r)["repo"],
+		id:    id,
+	}
```

Comment:
seems like these first ~10 lines get repeated a bunch in each function - maybe they can be generalized? (e.g. a function that returns the value of `key` and an `err`?) doesn't save too much line wise but would make it more foolproof later on :D

Reply:
niiiiiiice